metronome/hotstuff/service/test/src/io/iohk/metronome/hotstuff/service/execution/BlockExecutorProps.scala

```diff
@@ -0,0 +1,287 @@
+            blockHash
+          }
+        }
+
+    val lastBatchCommitedBlockHash =
+      batches.last.quorumCertificate.blockHash
+
+    def awaitBlockExecution(
+        blockHash: TestAgreement.Hash
+    ): Task[Vector[TestAgreement.Hash]] = {
+      executedBlockHashes
+        .restartUntil { blockHashes =>
+          blockHashes.contains(blockHash)
+        }
+    }
+  }
+
+  object TestFixture {
+    implicit val arb: Arbitrary[TestFixture] = Arbitrary(gen())
+
+    /** Create a random number of tree extensions, with each extension
+      * covered by a batch that goes from its root to one of its leaves.
+      */
+    def gen(minBatches: Int = 1, maxBatches: Int = 5): Gen[TestFixture] = {
+      def loop(
+          i: Int,
+          tree: List[TestBlock],
+          effects: Vector[Effect.ExecuteBlocks[TestAgreement]]
+      ): Gen[TestFixture] = {
+        if (i == 0) {
+          Gen.const(TestFixture(tree, effects))
+        } else {
+          val extension = for {
+            viewNumber <- Gen.posNum[Int].map(ViewNumber(_))
+            ancestor = tree.last
+            descendantTree <- genNonEmptyBlockTree(parentId = ancestor.id)
+            descendant = descendantTree.last
+            commitQC = QuorumCertificate[TestAgreement](
+              phase = Phase.Commit,
+              viewNumber = viewNumber,
+              blockHash = descendant.id,
+              signature = GroupSignature(())
+            )
+            effect = Effect.ExecuteBlocks[TestAgreement](
+              lastExecutedBlockHash = ancestor.id,
+              quorumCertificate = commitQC
+            )
+          } yield (tree ++ descendantTree, effects :+ effect)
+
+          extension.flatMap { case (tree, effects) =>
+            loop(i - 1, tree, effects)
+          }
+        }
+      }
+
+      for {
+        prefixTree <- genNonEmptyBlockTree
+        i          <- Gen.choose(minBatches, maxBatches)
+        fixture    <- loop(i, prefixTree, Vector.empty)
+      } yield fixture
+    }
+  }
+
+  def run(test: Task[Prop]): Prop = {
+    import Scheduler.Implicits.global
+    test.runSyncUnsafe(timeout = 5.seconds)
+  }
+
+  property("executeBlocks - from root") = forAll { (fixture: TestFixture) =>
+    run {
+      fixture.resources.use { case (blockSychronizer, _) =>
+        for {
+          _ <- fixture.batches.traverse(blockSychronizer.enqueue)
+
+          executedBlockHashes <- fixture.awaitBlockExecution(
+            fixture.lastBatchCommitedBlockHash
+          )
+
+          // The genesis was the only block we marked as executed.
+          pathFromRoot <- fixture.storeRunner.runReadOnly {
+            TestBlockStorage.getPathFromRoot(fixture.lastBatchCommitedBlockHash)
+          }
+
+        } yield {
+          "executes from the root" |: executedBlockHashes == pathFromRoot.tail
+        }
+      }
+    }
+  }
+
+  property("executeBlocks - from last") = forAll { (fixture: TestFixture) =>
```

Comment:
Should we set `maxBatches = 1` since we're only enqueuing the last one?

Reply:
Sorry, but I fail to see how the other batches are relevant here:

> at the time it takes on the next batch

```
_ <- blockSychronizer.enqueue(lastBatch)
```

Since we're executing a single batch, there is no next batch. Am I missing something?

Reply:
No, you're not missing anything, it's only using the last batch, the others may only be there to emphasise that their presence should make no difference. 

Sorry if I didn't make myself clear: I agree with you that setting `maxBatches = 1` would achieve the same effect. I just didn't want to specify something that doesn't have any bearing on whether the test should work, to avoid making it look like that's a crucial setting (like it is when I set `minBatches = 2` in the failure case). The only effect should be that test generation may become quicker.

Now, if there wasn't the `prefixTree` in the data generation (a 0th batch if you will), then it could have an effect, because in that case `maxBatches = 1` would mean always going from the root.

Reply:
> test generation may become quicker.

That's what I was after

> avoid making it look like that's a crucial setting

but I'm fine with this reasoning